Review of rdkit.Chem.Fraggle code

gedeck opened the pull request:
Improved test coverage to 100%
Use more pythonic idioms

greglandrum commented on rdkit/Chem/Fraggle/FraggleSim.py:
```diff
@@ -29,356 +29,285 @@
+from itertools import combinations
 
 # our default rdkit fingerprinter parameters:
-rdkitFpParams = dict(maxPath=5, fpSize=1024, nBitsPerHash=2)
```
Curious why you made this change.
Both forms are legit Python and I find the first more readable. 
I think I remember reading about a performance difference, but that's hardly relevant here.

gedeck replied:
Here is the link: https://doughellmann.com/blog/2012/11/12/the-performance-impact-of-using-dict-instead-of-in-cpython-2-7-2/

While I'm usually not a big fan of micro-optimisations, I started to switch to the {} notation and use this now consistently.

greglandrum commented on rdkit/Chem/Fraggle/FraggleSim.py:
```diff
@@ -29,356 +29,285 @@
-  #use the same parent mol object and create editable mol
+  # Replace the given bonds with attachment points (B1-B2 -> B1-[*].[*]-B2)
+  # print("Breaking bonds between atoms: ", bonds)
   em = Chem.EditableMol(mol)
```
Should be using Chem.RWMol now.

gedeck replied:
Done

greglandrum commented on rdkit/Chem/Fraggle/FraggleSim.py:
```diff
@@ -29,356 +29,285 @@
+    return True, m.GetNumAtoms()
+  else:
+    return False, 0
 
```
I guess you can probably just return the number of atoms here (or 0 for a bad molecule); the first element of the tuple is redundant

gedeck replied:
I changed the code. The initial intention was probably to return a flag (is_valid_ringcut) and the number of atoms was an after thought. I changed this now so that the function only does what the name implies and move the calculation of number of atoms to the calling code.

greglandrum commented on rdkit/Chem/Fraggle/FraggleSim.py:
```diff
@@ -29,356 +29,285 @@
+        # fragments where we only chop off a small part of the input cmpd
+        if fhac > 3:
+          result.append(f)
           result_hcount = result_hcount + fhac
```
As long as we're cleaning up, might as well switch this to +=

gedeck replied:
Changed.

greglandrum commented on rdkit/Chem/Fraggle/FraggleSim.py:
```diff
@@ -29,356 +29,285 @@
+    # need to find the fragments which are valid which means they must be:
     #  Terminal (one attachment point) or valid ring cut
     for f in fragments:
       attachments = f.count("*")
```
Minor point, but it feels like this should be named nAttachments or attachmentsCount or something like that

greglandrum commented on rdkit/Chem/Fraggle/FraggleSim.py:
```diff
@@ -29,356 +29,285 @@
-      if (attachments == 1):
+      # check if terminal fragment
+      if f.count("*") == 1:
         fMol = Chem.MolFromSmiles(f)
```
Performance thing for here and the other examples where MolFromSmiles is called to calculate the number of heavy atoms: sanitization is not necessary for that calculation, so it should be skipped.

gedeck replied:
Changed it.

greglandrum commented on rdkit/Chem/Fraggle/FraggleSim.py:
```diff
@@ -29,356 +29,285 @@
 
-def generate_fraggle_fragmentation(mol):
-  """
-    >>> q = Chem.MolFromSmiles('COc1cc(CN2CCC(NC(=O)c3cncc(C)c3)CC2)c(OC)c2ccccc12')
```
I would prefer to leave this documentation in here because it makes it very clear what the return looks like.

We can argue about whether or not we run the doctests, but that's a separate point

gedeck replied:
I added it back in.

greglandrum commented on rdkit/Chem/Fraggle/FraggleSim.py:
```diff
@@ -29,356 +29,285 @@
+  # loop to generate every single and double cut in the molecule
+  # single cuts are not required as relevant single cut fragments can be found
+  # from the double cuts. For explanation see check_fragments method
+  for bond1, bond2 in combinations(acyclic_matching_atoms, 2):
```
Nice. Itertools is great

greglandrum commented on rdkit/Chem/Fraggle/FraggleSim.py:
```diff
@@ -29,356 +29,285 @@
   subsFp = Chem.RDKFingerprint(qsMol, **rdkitFpParams)
 
-  #loop through atoms of smiles and mark
+  # loop through atoms of smiles get atoms that have a high similarity with substructure
```
Note to self: made it this far

greglandrum commented on rdkit/Chem/Fraggle/FraggleSim.py:
```diff
@@ -29,356 +29,285 @@
-    except Exception:
+    except Exception:  # pragma: nocover
       sys.stderr.write("Can't parse smiles: %s\n" % (Chem.MolToSmiles(pMol)))
       pMol = Chem.Mol(mol.ToBinary())
```
This can now just be pMol = Chem.Mol(mol)

gedeck replied:
Changed

greglandrum commented on rdkit/Chem/Fraggle/UnitTestFraggle.py:
```diff
@@ -6,24 +6,28 @@
 
 
-class TestCase(unittest.TestCase):
+def load_tests(loader, tests, ignore):  # pylint: disable=unused-argument
```
If I understand this correctly it is loading the doctests from FraggleSim.py and they will be executed. If that's the case, then FraggleSim.py should be removed from test_list.py

greglandrum replied:
yeah, please remove them from test_list.py.
I'm fine with them being brought in through the unit tests, but I don't want them to be run twice.

gedeck commented:
Added a few more changes now:

1. Use `Chem.FragmentOnBonds` in delete_bonds
2. Use `Chem.GetMolFrags` to split molecule into a list of molecules
3. Change `select_fragments` to take a list of molecules
4. Changed `is_ring_cut_valid` to `isValidRingCut` that takes now a mol and only checks against the SMARTS

The changes reduce the number of C-code calls which might translate into better performance.

greglandrum commented on rdkit/Chem/Fraggle/FraggleSim.py:
```diff
@@ -147,48 +125,44 @@ def select_fragments(f_smi, ftype, hac):
+    result_hcount = 0
+    for fMol in fragments:
+      f = Chem.MolToSmiles(fMol)
+      nAttachments = f.count("*")
```
Here's a much faster way to count dummy atoms:
```
from rdkit.Chem import rdqueries
qa = rdqueries.AtomNumEqualsQueryAtom(0);
nAttachments = len(mol.GetAtomsMatchingQuery(qa))
```

greglandrum commented on rdkit/Chem/Fraggle/FraggleSim.py:
```diff
@@ -197,8 +171,39 @@ def select_fragments(f_smi, ftype, hac):
+
+  For valid ring cuts, the function returns the number of atoms. Otherwise 0. """
+  # At this point, the molecule requires the identification of rings, so we need to sanitize
+  Chem.SanitizeMol(mol, Chem.SanitizeFlags.SANITIZE_SYMMRINGS)
```
If the only sanitize operation you need is the ring finding, you could just call the ring finding code directly.

gedeck commented:
- use rdqueries.AtomNumEqualsQueryAtom(0) to determine the number of attachment points

greglandrum commented:
I think this is ready to go. @gedeck : ok for me to merge now?

gedeck commented:
@greglandrum yes. no more changes from my side for now.
